### PR TITLE
[hail] Improve split_multi_hts performance

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

**Comment by tpoterba** [comment]
master:
```
2019-09-03 07:04:50,033: INFO: [1/1] Running split_multi_hts...
2019-09-03 07:05:54,871: INFO:     burn in: 64.79s
2019-09-03 07:07:00,623: INFO:     run 1: 65.74s
2019-09-03 07:08:05,181: INFO:     run 2: 64.56s
2019-09-03 07:09:07,775: INFO:     run 3: 62.59s
```

PR:
```
2019-09-03 07:10:34,045: INFO: [1/1] Running split_multi_hts...
2019-09-03 07:11:16,374: INFO:     burn in: 42.33s
2019-09-03 07:11:56,085: INFO:     run 1: 39.71s
2019-09-03 07:12:34,916: INFO:     run 2: 38.83s
2019-09-03 07:13:14,087: INFO:     run 3: 39.17s
```

**Comment by tpoterba** [comment]
This is almost all coming from the Python change. the gqFromPL change has a negligible effect on performance, but still seems good to include.

**Comment by tpoterba** [comment]
more improvements (though less simple) to come.

File: hail/src/main/scala/is/hail/expr/ir/functions/GenotypeFunctions.scala [hunk 1]
```diff
@@ -27,7 +27,7 @@ object GenotypeFunctions extends RegistryFunctions {
           tPL.isElementDefined(region, pl, i).mux(
             Code._empty,
             Code._fatal("PL cannot have missing elements.")),
-          pli := region.loadInt(tPL.loadElement(region, pl, i)),
+          pli := region.loadInt(tPL.loadElement(region, pl, len, i)),
```
Comment (jigold) [inline hail/src/main/scala/is/hail/expr/ir/functions/GenotypeFunctions.scala]: Why were you able to change the signature of loadElement? `len` doesn't look like it's used anywhere

File: hail/python/hail/methods/statgen.py [hunk 2]
```diff
@@ -2234,11 +2234,14 @@ def split_multi_hts(ds, keep_star=False, left_aligned=False, vep_root='vep'):
...
                                 hl.min((hl.range(0, hl.triangle(split.old_alleles.length()))
                                         .filter(lambda j: hl.downcode(hl.unphased_diploid_gt_index_call(j),
-                                                                      split.a_index) == hl.unphased_diploid_gt_index_call(i)
+                                                                      split.a_index).unphased_diploid_gt_index() == i
                                                 ).map(lambda j: split.PL[j]))))))
-        update_entries_expression['PL'] = pl
         if 'GQ' in entry_fields:
-            update_entries_expression['GQ'] = hl.or_else(hl.gq_from_pl(pl), split.GQ)
+            pl_gq_struct = hl.rbind(pl, lambda pl: hl.struct(PL=pl, GQ=hl.gq_from_pl(pl)))
+            update_entries_expression['PL'] = pl_gq_struct['PL']
```
Comment (jigold) [inline hail/python/hail/methods/statgen.py]: Is there a reason you added `update_entries_expression['PL']` here?

File: hail/src/main/scala/is/hail/expr/ir/functions/GenotypeFunctions.scala (hunk 1 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/functions/GenotypeFunctions.scala]: the signature without `len` will load the length to calculate how many missing bits there are, and therefore where the elements start. This signature is faster.

File: hail/python/hail/methods/statgen.py (hunk 2 above)
Comment (tpoterba) [reply hail/python/hail/methods/statgen.py]: sorry, don't fully understand the question.

This change relies on the deduplication logic added to `StructExpression.annotate` that tries to figure out if a struct was exploded `annotate(**thing)` and pulls `thing` out in a `Let`.

The change will bind the `pl_gq_struct` in a `Let` if both appear.

**Pull request merged**